File: bindings/library/src/main/java/com/avsystem/anjay/impl/NativeInputContext.java
/*
 * Copyright 2020 AVSystem <avsystem@avsystem.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.avsystem.anjay.impl;

import com.avsystem.anjay.Anjay.Objlnk;
import com.avsystem.anjay.AnjayException;
import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;

public final class NativeInputContext implements AutoCloseable {
    private long self;

    private native void init(NativeInputContextPointer context);

    private native void cleanup();

    private static final class BytesContext {
        public boolean messageFinished;
        // NOTE: actually used, on C++ side.
        @SuppressWarnings("unused")
        public ByteBuffer slice;

        public int bytesRead;

        public BytesContext(ByteBuffer slice) {
            this.messageFinished = false;
            this.slice = slice;
            this.bytesRead = 0;
        }
    }

    private static final class StringContext {
        public String value;
    }

    private static final class BooleanContext {
        public boolean value;
    }

    private static final class IntegerContext {
        public int value;
    }

    private static final class LongContext {
        public long value;
    }

    private static final class FloatContext {
        public float value;
    }

    private static final class DoubleContext {
        public double value;
    }

    private static final class ObjlnkContext {
        public Objlnk value;
    }

    private native int anjayGetBytes(BytesContext bytesContext);

    private native int anjayGetString(StringContext stringContext);

    private native int anjayGetI32(IntegerContext intContext);

    private native int anjayGetI64(LongContext intContext);

    private native int anjayGetFloat(FloatContext floatContext);

    private native int anjayGetDouble(DoubleContext floatContext);

    private native int anjayGetBool(BooleanContext boolContext);

    private native int anjayGetObjlnk(ObjlnkContext objlnkContext);

    public NativeInputContext(NativeInputContextPointer context) {
        init(context);
    }

    public int getInt() {
        final IntegerContext context = new IntegerContext();
        final int result = anjayGetI32(context);
        if (result < 0) {
            throw new AnjayException(result, "anjay_get_i32() failed");
        }
        return context.value;
    }

    public long getLong() {
        final LongContext context = new LongContext();
        final int result = anjayGetI64(context);
        if (result < 0) {
            throw new AnjayException(result, "anjay_get_i64() failed");
        }
        return context.value;
    }

    public float getFloat() {
        final FloatContext context = new FloatContext();
        final int result = anjayGetFloat(context);
        if (result < 0) {
            throw new AnjayException(result, "anjay_get_float() failed");
        }
        return context.value;
    }

    public double getDouble() {
        final DoubleContext context = new DoubleContext();
        final int result = anjayGetDouble(context);
        if (result < 0) {
            throw new AnjayException(result, "anjay_get_double() failed");
        }
        return context.value;
    }

    public boolean getBoolean() {
        final BooleanContext context = new BooleanContext();
        final int result = anjayGetBool(context);
        if (result < 0) {
            throw new AnjayException(result, "anjay_get_bool() failed");
        }
        return context.value;
    }

    public String getString() {
        final StringContext context = new StringContext();
        final int result = anjayGetString(context);
        if (result < 0) {
            throw new AnjayException(result, "anjay_get_string() failed");
        }
        return context.value;
    }

    public Objlnk getObjlnk() {
        final ObjlnkContext context = new ObjlnkContext();
        final int result = anjayGetObjlnk(context);
        if (result < 0) {
            throw new AnjayException(result, "anjay_get_objlnk() failed");
        }
        return context.value;
    }

    public boolean getBytes(ByteBuffer out) {
        final int position = out.position();
        final BytesContext context = new BytesContext(out.slice());
        int result = anjayGetBytes(context);
        if (result < 0) {
            throw new AnjayException(result, "anjay_get_bytes() failed");
        }
        out.position(position + context.bytesRead);
        return context.messageFinished;
    }

    public byte[] getAllBytes() {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        ByteBuffer buf = ByteBuffer.allocate(4096);

        boolean finished = false;
        while (!finished) {
            buf.clear();
            finished = this.getBytes(buf);
            output.write(buf.array(), 0, buf.position());
        }

        return output.toByteArray();
    }

    @Override
    public void close() {
        this.cleanup();
    }
}

File: native-library/src/native_input_context.cpp
/*
 * Copyright 2020-2021 AVSystem <avsystem@avsystem.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <anjay/anjay.h>

#include "./jni_wrapper.hpp"

#include "./native_input_context.hpp"

#include "./util_classes/accessor_base.hpp"
#include "./util_classes/byte_buffer.hpp"

NativeInputContext::NativeInputContext(
        jni::JNIEnv &env,
        jni::Object<utils::NativeInputContextPointer> &context)
        : ctx_(utils::NativeInputContextPointer::into_native(env, context)) {}

jni::jint
NativeInputContext::get_i32(jni::JNIEnv &env,
                            jni::Object<details::InputCtx<int32_t>> &ctx) {
    return get_value<int32_t>(env, ctx, &anjay_get_i32);
}

jni::jint
NativeInputContext::get_i64(jni::JNIEnv &env,
                            jni::Object<details::InputCtx<int64_t>> &ctx) {
    return get_value<int64_t>(env, ctx, &anjay_get_i64);
}

jni::jint
NativeInputContext::get_bool(jni::JNIEnv &env,
                             jni::Object<details::InputCtx<bool>> &ctx) {
    return get_value<bool>(env, ctx, &anjay_get_bool);
}

jni::jint
NativeInputContext::get_float(jni::JNIEnv &env,
                              jni::Object<details::InputCtx<float>> &ctx) {
    return get_value<float>(env, ctx, &anjay_get_float);
}

jni::jint
NativeInputContext::get_double(jni::JNIEnv &env,
                               jni::Object<details::InputCtx<double>> &ctx) {
    return get_value<double>(env, ctx, &anjay_get_double);
}

jni::jint NativeInputContext::get_string(
        jni::JNIEnv &env, jni::Object<details::InputCtx<std::string>> &ctx) {
    return get_value<std::string>(env, ctx, [&](auto *ctx, auto *out_value) {
        int result;
        do {
            char chunk[1024];
            result = anjay_get_string(ctx, chunk, sizeof(chunk));
            if (result >= 0) {
                out_value->append(chunk);
            }
        } while (result == ANJAY_BUFFER_TOO_SHORT);
        return result;
    });
}

jni::jint NativeInputContext::get_objlnk(
        jni::JNIEnv &env, jni::Object<details::InputCtx<utils::Objlnk>> &ctx) {
    return get_value<utils::Objlnk>(env, ctx, [&](auto *ctx, auto *out_value) {
        return anjay_get_objlnk(ctx, &out_value->oid, &out_value->iid);
    });
}

jni::jint
NativeInputContext::get_bytes(jni::JNIEnv &env,
                              jni::Object<details::InputCtx<uint8_t[]>> &ctx) {
    auto context_accessor =
            utils::AccessorBase<details::InputCtx<uint8_t[]>>{ env, ctx };
    auto slice = utils::ByteBuffer(
            env, context_accessor.get_value<jni::Object<utils::ByteBuffer>>(
                         "slice"));
    size_t capacity = slice.capacity();
    size_t bytes_read = 0;
    bool message_finished = false;
    std::vector<jni::jbyte> buffer(
            std::min(capacity, static_cast<size_t>(4096)));

    while (capacity > 0 && !message_finished) {
        size_t chunk_bytes_read;
        int result = anjay_get_bytes(ctx_, &chunk_bytes_read, &message_finished,
                                     buffer.data(),
                                     std::min(buffer.size(), capacity));
        if (result < 0) {
            return result;
        }
        capacity -= chunk_bytes_read;
        bytes_read += chunk_bytes_read;
        slice.put(buffer);
    }
    context_accessor.set_value<int>("bytesRead", static_cast<int>(bytes_read));
    context_accessor.set_value<bool>("messageFinished", message_finished);
    return 0;
}

void NativeInputContext::register_native(jni::JNIEnv &env) {
#define METHOD(MethodPtr, name) \
    jni::MakeNativePeerMethod<decltype(MethodPtr), (MethodPtr)>(name)

    // clang-format off
    jni::RegisterNativePeer<NativeInputContext>(
            env, jni::Class<NativeInputContext>::Find(env), "self",
            jni::MakePeer<NativeInputContext, jni::Object<utils::NativeInputContextPointer> &>,
            "init",
            "cleanup",
            METHOD(&NativeInputContext::get_i32, "anjayGetI32"),
            METHOD(&NativeInputContext::get_i64, "anjayGetI64"),
            METHOD(&NativeInputContext::get_bool, "anjayGetBool"),
            METHOD(&NativeInputContext::get_float, "anjayGetFloat"),
            METHOD(&NativeInputContext::get_double, "anjayGetDouble"),
            METHOD(&NativeInputContext::get_string, "anjayGetString"),
            METHOD(&NativeInputContext::get_objlnk, "anjayGetObjlnk"),
            METHOD(&NativeInputContext::get_bytes, "anjayGetBytes")
    );
    // clang-format on
}
